rfd/0136-modern-signature-algorithms.md

```diff
@@ -0,0 +1,881 @@
+* crypto/x509 uses SHA256 hash by default with all RSA public keys
+  (but this can be overridden)
+  <https://github.com/golang/go/blob/dbf9bf2c39116f1330002ebba8f8870b96645d87/src/crypto/x509/x509.go#L1411-L1414>
+* ssh only supports the PKCS#1 v1.5 signature scheme with RSA keys
+  <https://datatracker.ietf.org/doc/html/rfc8332>
+* FIPS 186-5 approves all listed options
+* BoringCrypto supports all listed options
+* We could consider PSS signatures instead of PKCS#1 v1.5 for TLS and JWT
+  signatures, but SSH does not support it.
+
+#### ECDSA
+
+Curves: P-256
+
+Digest/hash algorithms: SHA256
+
+Considerations:
+
+* ECDSA has good support across SSH and TLS protocols for both client and CA
+  certs.
+* ECDSA certs are supported by web browsers.
+* ECDSA key generation is *much* faster than RSA key generation.
+* ECDSA signatures are faster than RSA signatures.
+* FIPS 186-5 approves all listed options
+* BoringCrypto supports all listed options
+* The P-256 curve is the most popular, it is considered to be secure, and it has
+  the broadest support among external tools.
+* We could consider supporting the P-384 and P-521 curves for CAs.
+
+#### EdDSA
+
+Curves: Ed25519
+
+Digest/hash algorithms: none (the full message is signed without hashing)
+
+Considerations:
+
+* There is widespread support for Ed25519 SSH certs.
+* Go libraries support Ed25519 for TLS
+* Support for Ed25519 is *not* widespread in the TLS ecosystem.
+* YubiHSM and GCP KMS do *not* support Ed25519 keys.
+* Ed25519 is considered by some to be the fastest, most secure, most modern
+  option for SSH certs.
+* Ed25519 key generation is *much* faster than RSA key generation.
+* Ed25519 signatures are faster than RSA signatures.
+* FIPS 186-5 approves Ed25519
+* BoringCrypto does not support Ed25519
+  <https://go.googlesource.com/go/+/dev.boringcrypto/src/crypto/tls/boring.go#80>
+* Ed25519 is the only EdDSA curve supported in the Go standard library.
+
+#### Algorithms Summary
+
+* We are probably forced to continue unconditionally using RSA for database
+  certs, I'm assuming this would apply to both client and CA.
+* Ed25519 is a modern favourite for SSH, but TLS (and HSM, KMS) support is lacking.
+* Teleport CAs use separate keypairs for SSH and TLS, they do not need to use
+  the same algorithm.
+* Teleport derives client SSH and TLS certs from the same client keypair,
+  supporting different algorithms for each will require larger changes.
+* It seems it is time to split client SSH and TLS keys to support the popular
+  and secure Ed25519 algorithm for SSH and the widely-suported
+  `ECDSA_P256_SHA256` algorithm for TLS. This will also allows to evolve the
+  algorithms used for each protocol independently in the future.
+
+### FIPS
+
+When compiled in FIPS mode Teleport uses Go's BoringCrypto library, which does
+not yet support Ed25519.
+Therefore we must have a different set of `recommended` algorithms for FIPS mode:
```

Comment:
What would be the path to convert a cluster to FIPS mode, in case a customer wanted to do so? Right now all they have to do is install a fips-mode control plane and everything should just keep working as expected, but with the use of ed25519 SSH certs almost everything in the cluster is going to need rekeying. Can we just assume that customers in need of FIPS 140-3 cryptography will know that from the very beginning?

Reply:
That is certainly one argument against using Ed25519 by default before BoringCrypto gets support for it

Reply:
Added a mention of the migration path. It will be a bit painful for a few users who need to migrate existing clusters, but I don't think we can justify sticking with only FIPS-compliant algorithms for everyone forever